tests/bluetooth/tester/src/btp_bap.c

```diff
@@ -17,35 +18,1258 @@
+	if (audio_conn->end_points_count > 0) {
+		return;
+	}
+
+	bt_conn_get_info(conn, &conn_info);
+
+	if (conn_info.role == BT_HCI_ROLE_CENTRAL) {
+		discover_ases(audio_conn);
+	}
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed,
+};
+
+static struct bt_pacs_cap cap_sink = {
+	.codec = &default_codec,
+};
+
+static struct bt_pacs_cap cap_source = {
+	.codec = &default_codec,
+};
+
+static uint8_t ascs_supported_commands(const void *cmd, uint16_t cmd_len,
+				       void *rsp, uint16_t *rsp_len)
+{
+	struct btp_ascs_read_supported_commands_rp *rp = rsp;
+
+	/* octet 0 */
+	tester_set_bit(rp->data, BTP_ASCS_READ_SUPPORTED_COMMANDS);
+
+	*rsp_len = sizeof(*rp) + 1;
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static int server_stream_config(struct bt_conn *conn, struct bt_bap_stream *stream,
+				struct bt_codec *codec, struct bt_codec_qos_pref *qos)
+{
+	int err;
+	struct bt_bap_ep *ep;
+
+	err = bt_bap_unicast_server_config_ase(conn, stream, codec, qos);
+
+	if (err != 0) {
+		return err;
+	}
+
+	print_codec(&default_codec);
+
+	ep = stream->ep;
+	LOG_DBG("ASE Codec Config: ase_id %u dir %u", ep->status.id, ep->dir);
+	LOG_DBG("ASE Codec Config stream %p", stream);
+
+	btp_send_ascs_operation_completed_ev(conn, ep->status.id, BT_ASCS_CONFIG_OP, 0);
+
+	return 0;
+}
+
+static int client_create_unicast_group(struct audio_connection *audio_conn, uint8_t ase_id)
+{
+	int err;
+	struct bt_bap_unicast_group_stream_pair_param pair_params[MAX_STREAMS_COUNT];
+	struct bt_bap_unicast_group_stream_param stream_params[MAX_STREAMS_COUNT];
+	struct bt_bap_unicast_group_param param;
+	size_t stream_cnt = 0;
 
-#define AVAILABLE_SINK_CONTEXT SUPPORTED_SINK_CONTEXT
-#define AVAILABLE_SOURCE_CONTEXT SUPPORTED_SOURCE_CONTEXT
+	for (size_t i = 0; i < MAX_STREAMS_COUNT; i++) {
+		struct bt_bap_stream *stream = &audio_conn->streams[i].stream;
 
-static struct bt_codec lc3_codec =
-	BT_CODEC_LC3(BT_CODEC_LC3_FREQ_ANY, BT_CODEC_LC3_DURATION_10,
-		     BT_CODEC_LC3_CHAN_COUNT_SUPPORT(1),
-		     40u, 120u, 1u,
-		     (BT_AUDIO_CONTEXT_TYPE_CONVERSATIONAL |
-		      BT_AUDIO_CONTEXT_TYPE_MEDIA));
+		if (stream == NULL || stream->ep == NULL) {
+			continue;
+		}
 
-static struct bt_pacs_cap cap_sink = {
-	.codec = &lc3_codec,
-};
+		if (stream->ep->status.id != ase_id) {
+			/* TODO: For now one stream per group is configured */
+			continue;
+		}
 
-static struct bt_pacs_cap cap_source = {
-	.codec = &lc3_codec,
+		stream_params[stream_cnt].stream = stream;
+		stream_params[stream_cnt].qos = &audio_conn->qos;
+
+		if (stream->ep->dir == BT_AUDIO_DIR_SOURCE) {
+			pair_params[stream_cnt].tx_param = NULL;
+			pair_params[stream_cnt].rx_param = &stream_params[stream_cnt];
+		} else {
+			pair_params[stream_cnt].tx_param = &stream_params[stream_cnt];
+			pair_params[stream_cnt].rx_param = NULL;
+		}
+
+		stream_cnt++;
+	}
+
+	if (stream_cnt == 0) {
+		return -1;
```

Comment:
For all places where you return `-1` I suggest to return an `errno` value instead